torch/distributed/optim/named_optimizer.py

```diff
@@ -0,0 +1,237 @@
+                "initialize the optimizer, not all parameters of the module."
+            )
+            param_to_key = {param: key for key, param in self.named_parameters}  # type: ignore[misc, has-type]
+            param_keys_order = []
+            for group in param_groups:
+                for param in group["params"]:
+                    if param not in param_to_key:
+                        raise ValueError("Param name not found for param group.")
+                    param_keys_order.append(param_to_key[param])
+            self.param_keys_order = param_keys_order
+
+    def state_dict(self) -> Dict[str, Any]:
+        """
+        Return the state_dict of the optimzer. Instead of using number to index
+        parameters, we will use module fully qualifed name (FQN) as the key.
+        """
+        state_dict = self._optimizer.state_dict()
+        param_groups = state_dict["param_groups"]
+
+        ret_state = {
+            self.param_keys_order[st_key]: state_val
+            for st_key, state_val in state_dict["state"].items()
+        }
+
+        ret_groups = []
+        for group in param_groups:
+            param_keys = []
+            for param in group["params"]:
+                param_keys.append(self.param_keys_order[param])
+            ret_group = {"params": sorted(param_keys)}
+            for k, v in group.items():
+                if k != "params":
+                    ret_group[k] = deepcopy(v)
+            ret_groups.append(ret_group)
+
+        ret: Dict[str, object] = {"state": ret_state}
+        ret["param_groups"] = ret_groups
+        return ret
+
+    def register_state_dict_pre(self):
+        raise NotImplementedError()
+
+    def register_state_dict_post(self):
+        raise NotImplementedError()
+
+    def step(self):
+        """
+        Performs a single optimization step.
+
+        This will call :meth:`torch.optim.Optimizer.step` on the wrapped
+        optimizer.
+        """
+        self._optimizer.step()
+
+    def load_state_dict(self, state_dict: Mapping[str, Any]) -> None:
+        """
+        This public function defines the default behavior to load a state_dict
+        for named optimizer.
+
+        Sample Code
+        ```
+            my_model = MyModule()
+            optimizer = NamedOptimizer(my_model.named_parameters(), Adagrad)
+            ...
+
+            optim_state_dict = optimizer.state_dict()
+            ...
+
+            fs_storage_loader = torch.distributed.FileSystemLoader("/checkpoint/1")
+            torch.distributed.load_state_dict(
+                state_dict=optim_state_dict,
+                storage_reader=fs_storage_loader,
+            )
+
+            optim_state_dict.load_state_dict(optim_state_dict)
+            ...
+        ```
+        Args:
+            state_dict (Dict[str, Any]) : A state_dict to load to. Note that this
+                state dict will updated in places.
+        """
+        new_state_dict = self._optimizer.state_dict()
```

Comment:
IIUC, if the user calls `NamedOptimizer.load_state_dict(state_dict)` but the wrapped optimizer's state has not been initialized yet, then this `NamedOptimizer.load_state_dict()` method will error somewhere while trying to load (looks like L162 if `len(new_state) != len(state)`).

Does this match the behavior of `torch.optim.Optimizer.load_state_dict()`?

Reply:
That's a good point, `torch.optim.Optimizer.load_state_dict()` does allow this behavior but keyedOptimizer in TorchRec does not. I don't know what we want from NamedOptimizer for this behavior and I agree we need to handle this case. Let me add a TODO here. This PR it too big, I don't want to include it here.

Reply:
I belive TorchRec optimizer provides an API to eagerly initialize the optimizer state.